### PR TITLE
feat: add Jina Reranking integration

### PR DESCRIPTION
**Pull request by JoanFM**
# Description

This PR aims to add support for Jina AI `reranking` the same way it does for `embeddings`.

## Type of Change

Please delete options that are not relevant.

- [x] New feature (non-breaking change which adds functionality)
- [x] This change requires a documentation update

# How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

- [ ] Added new unit/integration tests
- [ ] Added new notebook (that tests end-to-end)
- [ ] I stared at the code and made sure it makes sense

# Suggested Checklist:

- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have added Google Colab support for the newly added notebooks.
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [x] I ran `make format; make lint` to appease the lint gods

### CODE REVIEW

File: llama-index-integrations/postprocessor/llama-index-postprocessor-jinaai-rerank/pyproject.toml [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+check-filenames = true
+check-hidden = true
+skip = "*.csv,*.html,*.json,*.jsonl,*.pdf,*.txt,*.ipynb"
+
+[tool.llamahub]
+contains_example = false
+import_path = "llama_index.postprocessor.jinaai_rerank"
+
+[tool.llamahub.class_authors]
+JinaRerank = "llama-index"
```
Comment (nerdai) [inline llama-index-integrations/postprocessor/llama-index-postprocessor-jinaai-rerank/pyproject.toml:15]: feel free to add your github username here if you like!

Comment (JoanFM) [reply llama-index-integrations/postprocessor/llama-index-postprocessor-jinaai-rerank/pyproject.toml]: done, thanks

File: llama-index-integrations/postprocessor/llama-index-postprocessor-jinaai-rerank/README.md [hunk 2]
```diff
@@ -0,0 +1 @@
+# LlamaIndex Postprocessor Integration: Jina AI Rerank
```
Comment (hatianzhang) [inline llama-index-integrations/postprocessor/llama-index-postprocessor-jinaai-rerank/README.md:1]: pls provide more context for your reranker here

File: llama-index-integrations/postprocessor/llama-index-postprocessor-jinaai-rerank/llama_index/postprocessor/jinaai_rerank/base.py [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+                raise RuntimeError(resp["detail"])
+
+            results = resp["results"]
+
+            new_nodes = []
+            for result in results:
+                new_node_with_score = NodeWithScore(
+                    node=nodes[result["index"]].node, score=result["relevance_score"]
+                )
+                new_nodes.append(new_node_with_score)
```
Comment (hatianzhang) [inline llama-index-integrations/postprocessor/llama-index-postprocessor-jinaai-rerank/llama_index/postprocessor/jinaai_rerank/base.py:80]: I think you need to sort based on the new score to overwrite previous rank list

**Comment by hatianzhang** [comment]
pls also add a notebook to showcase how to use it and your results

File: llama-index-integrations/postprocessor/llama-index-postprocessor-jinaai-rerank/llama_index/postprocessor/jinaai_rerank/base.py (hunk 3 above)
Comment (JoanFM) [reply llama-index-integrations/postprocessor/llama-index-postprocessor-jinaai-rerank/llama_index/postprocessor/jinaai_rerank/base.py:80]: the results are returned in order

**Comment by JoanFM** [comment]
> pls also add a notebook to showcase how to use it and your results

where should I put the notebook? and is there an example for me to follow?

**Comment by hatianzhang** [comment]
@JoanFM you could follow/put here https://github.com/run-llama/llama_index/blob/a0123bb18d930e9e5e5e8a5e9a908f8be47ff3bc/docs/examples/node_postprocessor/ColbertRerank.ipynb#L4

**Comment by JoanFM** [comment]
> Contributor

done!

File: docs/examples/node_postprocessor/JinaRerank.ipynb [hunk 4]
```diff
@@ -0,0 +1,173 @@
+{
```
Comment (hatianzhang) [inline docs/examples/node_postprocessor/JinaRerank.ipynb:1]: <p>could you print out the reranked nodes from response </p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/run-llama/llama_index/pull/11291/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='11'/>

**Review by hatianzhang — APPROVED** [review]

File: docs/examples/node_postprocessor/JinaRerank.ipynb (hunk 4 above)
Comment (JoanFM) [reply docs/examples/node_postprocessor/JinaRerank.ipynb:1]: how is this printed?

File: docs/examples/node_postprocessor/JinaRerank.ipynb [hunk 5]
```diff
@@ -0,0 +1,160 @@
...
```
Comment (hatianzhang) [inline docs/examples/node_postprocessor/JinaRerank.ipynb:1]: <p>the output block seems missing</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/run-llama/llama_index/pull/11291/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='12'/>

Comment (JoanFM) [reply docs/examples/node_postprocessor/JinaRerank.ipynb:1]: U want me to show the output of the model? I will

Comment (hatianzhang) [reply docs/examples/node_postprocessor/JinaRerank.ipynb:1]: <p>yes, better to show the output</p>

**Review by hatianzhang — APPROVED** [review]

**Comment by nerdai** [comment]
@JoanFM @hatianzhang: Thanks!

Published to pypi (see [here](https://pypi.org/project/llama-index-postprocessor-jinaai-rerank/))

**Pull request merged**